addons/spreadsheet/static/src/helpers/model.js

```diff
@@ -0,0 +1,167 @@
+/** @odoo-module */
+import { DataSources } from "@spreadsheet/data_sources/data_sources";
+import * as spreadsheet from "@odoo/o-spreadsheet";
+import { migrate } from "@spreadsheet/o_spreadsheet/migration";
+import { _t } from "@web/core/l10n/translation";
+import { loadJS } from "@web/core/assets";
+
+const { Model, tokenize } = spreadsheet;
+const { toCartesian } = spreadsheet.helpers;
+
+export async function fetchSpreadsheetModel(orm, resModel, resId) {
+    const { data, revisions } = await orm.call(resModel, "join_spreadsheet_session", [resId]);
+    return createSpreadsheetModel({ orm, data, revisions });
+}
+
+export function createSpreadsheetModel({ orm, data, revisions }) {
+    const dataSources = new DataSources(orm);
+    const model = new Model(migrate(data), { custom: { dataSources } }, revisions);
+    return model;
+}
+
+/**
+ * Ensure that the spreadsheet does not contains cells that are in loading state
+ * @param {Model} model
+ * @returns {Promise<void>}
+ */
+export async function waitForDataLoaded(model) {
+    const dataSources = model.config.custom.dataSources;
+    await dataSources.waitForAllLoaded();
+    return new Promise((resolve, reject) => {
+        function check() {
+            model.dispatch("EVALUATE_CELLS");
+            if (isLoaded(model)) {
+                dataSources.removeEventListener("data-source-updated", check);
+                resolve();
+            }
+        }
+        dataSources.addEventListener("data-source-updated", check);
+        check();
+    });
+}
+
+/**
+ * @param {Model} model
+ * @returns {object}
+ */
+export async function freezeOdooData(model) {
+    await waitForDataLoaded(model);
+    const data = model.exportData();
+    for (const sheet of Object.values(data.sheets)) {
+        for (const [xc, cell] of Object.entries(sheet.cells)) {
+            if (containsOdooFunction(cell.content)) {
+                const { col, row } = toCartesian(xc);
+                const sheetId = sheet.id;
+                const evaluatedCell = model.getters.getEvaluatedCell({
+                    sheetId,
+                    col,
+                    row,
+                });
+                cell.content = evaluatedCell.formattedValue;
+                if (evaluatedCell.format) {
+                    cell.format = getItemId(evaluatedCell.format, data.formats);
+                }
+            }
+        }
+        for (const figure of sheet.figures) {
+            if (figure.tag === "chart" && figure.data.type.startsWith("odoo_")) {
+                await loadJS("/web/static/lib/Chart/Chart.js");
+                const img = odooChartToImage(model, figure);
+                figure.tag = "image";
+                figure.data = {
+                    path: img,
+                    size: { width: figure.width, height: figure.height },
```

Comment:
there is a hiccup here,  before creating an image, you have to add it to the filestore, which in turn yields an attachment path. Said path will then be processed in `_zip_xslx_files` to actually bundle the image in the final xlsx file. here your blob will work when displaying a figure but won't be considered in the excel bundle.

I guess it could be addressed by instanciating a simple `RecordFileStore` though

Reply:
It should be addressed at the excel export level, not specifically when "freezing" the data